src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -1956,18 +1956,19 @@ private BoundExpression SynthesizeReceiver(SyntaxNode node, Symbol member, Bindi
                 (currentType.IsInterface && (declaringType.IsObjectType() || currentType.AllInterfacesNoUseSiteDiagnostics.Contains(declaringType))))
             {
                 bool hasErrors = false;
-                if (EnclosingNameofArgument != node)
+                if (!IsInsideNameof || (EnclosingNameofArgument != node && !IsFeatureAvailable(node, MessageID.IDS_FeatureReducedMemberAccessChecksInNameof)))
                 {
+                    var diagnosticsTemp = IsInsideNameof ? BindingDiagnosticBag.Discarded : diagnostics;
```

Comment:
>BindingDiagnosticBag.Discarded [](http://example.com/codeflow?start=59&length=30)

This looks suspicious. Even if we don't want to report any errors, we probably still want to track dependencies.

Reply:
I see that, given the current usage of ```diagnosticsTemp``` here and in ```IsRefOrOutThisParameterCaptured```, it is not possible to accumulate any dependencies. However, this still feels fragile. I also find it fragile to indiscriminately ignore all diagnostics that ```IsRefOrOutThisParameterCaptured``` might produce. 

What do you think about doing the following:
- extracting a helper from ```IsRefOrOutThisParameterCaptured``` that wouldn't take a diagnostic bag and would return a ```DiagnosticInfo``` or null instead of a bool.
- changing the code below to produce ```DiagnosticInfo```s instead of reporting diagnostic on the spot.
- conditionally reporting the ```DiagnosticInfo``` into ```diagnostics``` at the end (it looks like we use the same location for all errors).
- removing the ```diagnosticsTemp``` local completely.

It feels like, if we structure implementation this way, we will have clear guarantee that we are not dropping anything useful now and won't start dropping it in the future.